src/main/java/spiffe/provider/CertificateUtils.java

```diff
@@ -0,0 +1,176 @@
+    }
+
+    /**
+     * Generates a PrivateKey from the X509SVIDKey ByteArray
+     *
+     * It uses PKCS8EncodedKeySpec that represents the ASN.1 encoding of a private key
+     *
+     * @param input as byte array
+     * @return
+     * @throws NoSuchAlgorithmException
+     * @throws InvalidKeySpecException
+     */
+    static PrivateKey generatePrivateKey(byte[] input) throws NoSuchAlgorithmException, InvalidKeySpecException {
+        KeyFactory keyFactory = KeyFactory.getInstance("EC", new BouncyCastleProvider());
+        return keyFactory.generatePrivate(new PKCS8EncodedKeySpec(input));
+    }
+
+    /**
+     * Validate a certificate chain against a set of trusted certificates.
+     *
+     * @param chain    certificate chain
+     * @param trustedCerts
+     * @throws CertificateException
+     */
+    static void validate(X509Certificate[] chain, Set<X509Certificate> trustedCerts) throws CertificateException {
+        PKIXParameters pkixParameters = toPkixParameters(trustedCerts);
+        CertPath certPath = certificateFactory.generateCertPath(asList(chain));
+        try {
+            certPathValidator.validate(certPath, pkixParameters);
+        } catch (CertPathValidatorException | InvalidAlgorithmParameterException e) {
+            throw new CertificateException(e);
+        }
+    }
+
+    /**
+     * Validates that the SPIFFE ID is present and matches the SPIFFE ID configured in
+     * the java.security property ssl.spiffe.accept
+     *
+     * @param chain an array of X509Certificate that contains the Peer's SVID to be validated
+     * @throws CertificateException when either the certificates doesn't have a SPIFFE ID or the SPIFFE ID is not trusted
+     */
+    static void checkSpiffeId(X509Certificate[] chain) throws CertificateException {
+        Optional<String> spiffeId = getSpiffeId(chain[0]);
+        if (spiffeId.isPresent()) {
+            String acceptedSpiffeId = Security.getProperty("ssl.spiffe.accept");
+
+            if (!StringUtils.equals(spiffeId.get(), acceptedSpiffeId)) {
+                String errorMessage = format("SPIFFE ID %s is not a trusted", spiffeId.get());
+                LOGGER.log(Level.WARNING, errorMessage);
+                throw new CertificateException(errorMessage);
+            }
+        } else {
+            throw new CertificateException("SPIFFE ID not found in the certificate");
+        }
+    }
+
+
+    /**
+     * Extracts the SpiffeID from a SVID - X509Certificate
+     *
+     * @param certificate
+     * @return
+     * @throws CertificateParsingException
+     */
+    private static Optional<String> getSpiffeId(X509Certificate certificate) throws CertificateParsingException {
+        return certificate.getSubjectAlternativeNames().stream()
+                .map(san -> (String) san.get(1))
+                .filter(uri -> startsWith(uri, "spiffe://"))
+                .findFirst();
```

Comment:
what happens if there is more than 1? is it allow? must not we return an error in that case?

Reply:
done. Throwing an error when there're multiple spiffe ids in the certificate, that's not supported.